examples/pytorch/speech-recognition/run_speech_recognition_ctc.py

```diff
@@ -0,0 +1,585 @@
+        data_args.dataset_name, data_args.dataset_config_name, split=data_args.eval_split_name
+    )
+
+    if data_args.audio_column_name not in raw_datasets["train"].column_names:
+        raise ValueError(
+            f"--audio_column_name {data_args.audio_column_name} not found in dataset '{data_args.dataset_name}'. "
+            "Make sure to set `--audio_column_name` to the correct audio column - one of "
+            f"{', '.join(raw_datasets['train'].column_names)}."
+        )
+
+    if data_args.text_column_name not in raw_datasets["train"].column_names:
+        raise ValueError(
+            f"--text_column_name {data_args.audio_column_name} not found in dataset '{data_args.dataset_name}'. "
+            "Make sure to set `--text_column_name` to the correct text column - one of "
+            f"{', '.join(raw_datasets['train'].column_names)}."
+        )
+
+    # prepare dataset
+    if data_args.max_train_samples is not None:
+        raw_datasets["train"] = raw_datasets["train"].select(range(data_args.max_train_samples))
+
+    if data_args.max_eval_samples is not None:
+        raw_datasets["eval"] = raw_datasets["eval"].select(range(data_args.max_eval_samples))
+
+    # 2. We remove some special characters from the datasets
+    # that make training complicated and do not help in transcriping the speech
+    # E.g. characters, such as `,` and `.` do not really have an acoustic characteristic
+    # that could be easily picked up by the model
+
+    chars_to_ignore_regex = (
+        f'[{"".join(data_args.chars_to_ignore)}]' if data_args.chars_to_ignore is not None else None
+    )
+
+    def remove_special_characters(batch):
+        if chars_to_ignore_regex is not None:
+            batch["target_text"] = re.sub(chars_to_ignore_regex, "", batch[data_args.text_column_name]).lower() + " "
+        else:
+            batch["target_text"] = batch[data_args.text_column_name].lower() + " "
+        return batch
+
+    raw_datasets = raw_datasets.map(remove_special_characters, remove_columns=[data_args.text_column_name])
+
+    # 3. Next, we create the vocabulary of the model by extracting all unique characters from
+    # the training and evaluation datasets
+    # We need to make sure that only first rank saves vocabulary
+    if training_args.world_size == 1 or dist.get_rank() == 0:
+        vocab_dict = create_vocabulary_from_data(raw_datasets)
+
+        vocab_file = os.path.join(training_args.output_dir, "vocab.json")
+
+        # save vocab dict to be loaded into tokenizer
+        os.makedirs(training_args.output_dir, exist_ok=True)
+        if training_args.overwrite_output_dir and os.path.isfile(vocab_file):
+            os.remove(vocab_file)
+
+        if not os.path.isfile(vocab_file):
+            with open(vocab_file, "w") as vocab_file:
+                json.dump(vocab_dict, vocab_file)
+
+    # make sure all processes wait until vocab is created
+    if training_args.world_size > 1:
+        dist.barrier()
```

Comment:
this is important so that other processes don't error out when the first process hasn't saved the vocab yet

Reply:
Note this won't work on TPU or sagemaker. You should use the context manager `with training_args.main_process_first()` for the previous code to have something that works for any kind of distributed setup out of the box.

Reply:
That's a really nice context manager! Thanks for pointing it out